Mck/16649/trunk

ifesdjeen opened the pull request.

michaelsembwever committed:
Add Versions.Major.v4X to Java DTests (after cassandra-4.0 branch was created)
Change upgrades from a single path, to all upgrade paths that are globally supported within the specified from-to range.
Change all upgrades to v40 and v4X to be open ended (i.e. implicit CURRENT version).

 patch by Mick Semb Wever; reviewed by Alex Petrov for CASSANDRA-16649

michaelsembwever committed:
THROWAWAY when 0.0.8 released. in the meantime bump resolver to include apache snapshot repository

ifesdjeen commented on .build/build-resolver.xml:
```diff
@@ -56,6 +56,7 @@
         <resolver:remoterepos id="all">
             <remoterepo id="resolver-central" url="${artifact.remoteRepository.central}"/>
             <remoterepo id="resolver-apache" url="${artifact.remoteRepository.apache}"/>
+            <remoterepo id="resolver-apache-snapshots" url="${artifact.remoteRepository.apache-snaphot}" snapshots="true"/>
```
I vaguely remember someone being opposed to adding snapshots when I did the same, but tbh I can not recall the details.

michaelsembwever replied:
yes yes! that was me. this is in a separate commit and will be removed before the merge.

**OR** we find a way to allow the snapshot repository in between releases… more thought needed…

ifesdjeen commented on test/distributed/org/apache/cassandra/distributed/impl/InstanceConfig.java:
```diff
@@ -286,13 +286,12 @@ public static InstanceConfig generate(int nodeNum,
+        if (Versions.Major.v40.compareTo(major) <= 0)
+            return this;
+        else
+            return new InstanceConfig(this)
```
We could actually add a small comment on compatibility of `seeds` between the versions here. Definitely not a problem with a patch itself, just can use it as an opportunity to improve, while we have spotted it.

ifesdjeen commented on test/distributed/org/apache/cassandra/distributed/upgrade/MixedModeReplicationTestBase.java:
```diff
@@ -32,7 +33,12 @@
 public class MixedModeReplicationTestBase extends UpgradeTestBase
 {
-    protected void testSimpleStrategy(Versions.Major from, Versions.Major to) throws Throwable
+    protected void testSimpleStrategy(Versions.Major from) throws Throwable
```
Should we maybe keep `from` and `to`, and instead use `CURRENT` to indicate we'd like to upgrade to whatever the version we have right now? 

We had a plan to switch to an external repo for upgrade tests, and won't need to modify test base code when we do this.

michaelsembwever replied:
Done in `UpgradeTestBase.upgradeFrom(from)`

ifesdjeen commented on test/distributed/org/apache/cassandra/distributed/upgrade/UpgradeTestBase.java:
```diff
@@ -116,16 +129,24 @@ public TestCase nodes(int nodeCount)
+        public TestCase upgrade(Major from, @Nullable Major to)
+        {
+            SUPPORTED_UPGRADE_PATHS.stream()
+                .filter(upgradePath -> (upgradePath.left.compareTo(from) >= 0 && (null == to || upgradePath.right.compareTo(to) <= 0)))
```
If I understand this code correctly, we might want to have strict comparisons here: `upgradePath.left == to`. I'm assuming we're looking for the exact matches, otherwise by adding upgrade from one lower version to another we inadvertently allow other upgrades.

michaelsembwever replied:
The idea here is to add additional upgrades.

For example, if I define the test to be valid for upgrades from `v30`, and for upgrades upgrading up until `v4X`, that means we should be testing the upgrade paths:
 - `v30` -> `v3X`
 - `v30` -> `v40`
 - `v3X` -> `v40`
 - `v40` -> `v4X`
 
What if I rename
`upgrade(Major from)` to `allUpgradePathsFrom(from)`
and
`upgrade(Major from, Major to)` to `allUpgradePathsBetween(Major from, Major to)` ?

michaelsembwever replied:
I've renames both methods to `upgradesFrom(from)` and `upgrades(from,to)`
with the intentional that the plural verb helps indicate the change and difference.

michaelsembwever committed:
SQUASH fixes failed tests.
 - tests need a way to avoid specific upgrade paths (maybe it would be better to make assertions version-sensitive instead?)
 - rename `upgrade(Version initial, Version upgrade)` to `singleUpgrade(Major from, Major to)`
